operators/src/adapters/raster_time.rs

```diff
@@ -0,0 +1,379 @@
+use crate::engine::QueryRectangle;
+use crate::source::TilingStrategy;
+use crate::util::Result;
+use futures::stream::Zip;
+use futures::Stream;
+use futures::{ready, StreamExt};
+use geoengine_datatypes::primitives::{BoundingBox2D, TimeInstance, TimeInterval};
+use geoengine_datatypes::raster::{GridSize, Pixel, RasterTile2D, TileInformation};
+use pin_project::pin_project;
+use std::cmp::min;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// Merges two raster sources by aligning the temporal validity.
+/// Assumes that the raster tiles already align spatially.
+/// Assumes that raster tiles are contiguous temporally, with no-data-tiles filling gaps.
+/// Potentially queries the same tiles multiple times from its sources.
+#[pin_project(project = RasterTimeAdapterProjection)]
+pub struct RasterTimeAdapter<T1, T2, St1, St2, F1, F2>
+where
+    T1: Pixel,
+    T2: Pixel,
+    St1: Stream<Item = Result<RasterTile2D<T1>>>,
+    St2: Stream<Item = Result<RasterTile2D<T2>>>,
+    F1: Fn(QueryRectangle) -> St1,
+    F2: Fn(QueryRectangle) -> St2,
+{
+    source_a: F1,
+    source_b: F2,
+    query_rect: QueryRectangle,
+    time_end: Option<(TimeInstance, TimeInstance)>,
+    // TODO: calculate at start when tiling info is available before querying first tile
+    num_spatial_tiles: Option<usize>,
+    current_spatial_tile: usize,
+    #[pin]
+    stream: Zip<St1, St2>,
+    ended: bool,
+}
+
+impl<T1, T2, St1, St2, F1, F2> RasterTimeAdapter<T1, T2, St1, St2, F1, F2>
+where
+    T1: Pixel,
+    T2: Pixel,
+    St1: Stream<Item = Result<RasterTile2D<T1>>>,
+    St2: Stream<Item = Result<RasterTile2D<T2>>>,
+    F1: Fn(QueryRectangle) -> St1,
+    F2: Fn(QueryRectangle) -> St2,
+{
+    pub fn new(source_a: F1, source_b: F2, query_rect: QueryRectangle) -> Self {
+        Self {
+            stream: source_a(query_rect).zip(source_b(query_rect)),
+            source_a,
+            source_b,
+            query_rect,
+            num_spatial_tiles: None,
+            current_spatial_tile: 0,
+            time_end: None,
+            ended: false,
+        }
+    }
+
+    // TODO: move to TimeInterval
+    fn intersect_time_intervals(t1: TimeInterval, t2: TimeInterval) -> Option<TimeInterval> {
+        if t1.intersects(&t2) {
+            let start = std::cmp::max(t1.start, t2.start);
+            let end = std::cmp::min(t1.end, t2.end);
+            Some(TimeInterval::new_unchecked(start, end))
+        } else {
+            None
+        }
+    }
+
+    fn align_tiles(
+        mut tile_a: RasterTile2D<T1>,
+        mut tile_b: RasterTile2D<T2>,
+    ) -> (RasterTile2D<T1>, RasterTile2D<T2>) {
+        // TODO: scale data if measurement unit requires it?
+        let time = Self::intersect_time_intervals(tile_a.time, tile_b.time)
+            .expect("intervals must overlap");
```

Comment:
Looks like you should use a `Result` in the first place

Reply:
no, at the struct definition the documentation says;
> /// Assumes that the raster tiles already align spatially.
/// Assumes that raster tiles are contiguous temporally, with no-data-tiles filling gaps.

so the two tiles *must* overlap. Otherwise the source operators validate this assumption and I have to panic here